include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -157,32 +161,291 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+
+			CCache() {}
+
+			// TODO: add methods as needed, e.g. to serialize and deserialize to/from a pointer
+			std::vector<uint8_t> serialize();
+			static core::smart_refctd_ptr<CCache> deserialize(core::smart_refctd_ptr<ICPUBuffer> serializedCache);
+			static core::smart_refctd_ptr<CCache> deserialize(std::span<uint8_t> serializedCache);
+
+		private:
+			// we only do lookups based on main file contents + compiler options
+			struct Hash
+			{
+				inline size_t operator()(const SEntry& entry) const noexcept
+				{
+					std::vector<char> hashable = entry.compilerData.getHashable();
+					std::string hashableString(hashable.begin(), hashable.end());
+					hashableString += entry.mainFileContents;
+					return std::hash<std::string>{}(hashableString);
+				}
+				
+			};
+			struct KeyEqual
+			{
+				// used for insertions
+				inline bool operator()(const SEntry& lhs, const SEntry& rhs) const
+				{
+					return lhs.compilerData == rhs.compilerData && lhs.mainFileContents == rhs.mainFileContents;
+				}
+				
+			};
+
+			
+
+			// json serialization methods for the unordered multiset go here since I need access to Hash and KeyEquals 
+
+			core::unordered_multiset<SEntry, Hash, KeyEqual> m_container;
+			// If provided at creation time, this will hold the json representing entries in m_container + all SPIRV shaders
+			core::smart_refctd_ptr<ICPUBuffer> storageBuffer = nullptr;
+			// Specifies the size in bytes of the container json. This is so we know the offset into the buffer where shader bytecode starts
+			uint64_t containerJsonSize = 0;
 		};
 
+		inline core::smart_refctd_ptr<ICPUShader> compileToSPIRV(const std::string_view code, const SCompilerOptions& options, core::smart_refctd_ptr<CCache> cache = nullptr) const
+		{
+			/*core::vector<CCache::SEntry::SPreprocessingDependency> dependencies;
+			if (options.cache)
+			{
+				const auto& mainDep = dependencies.emplace_back(code,options.hash());
+				auto found = options.cache->find(mainDep);
+				if (found)
+					return core::smart_refctd_ptr<ICPUShader>(found);
+			}
+			auto retval = compileToSPIRV_impl(code,options,dependencies);
+			if (options.cache)
+				options.cache->insert(CCache::SEntry(dependencies));
+			return retval;*/
+			return compileToSPIRV_impl(code, options, nullptr);
+		}
 
-		virtual core::smart_refctd_ptr<ICPUShader> compileToSPIRV(const char* code, const SCompilerOptions& options) const = 0;
+		inline core::smart_refctd_ptr<ICPUShader> compileToSPIRV(const char* code, const SCompilerOptions& options, core::smart_refctd_ptr<CCache> cache = nullptr) const
+		{
+			if (!code)
+				return nullptr;
+			return compileToSPIRV({code,strlen(code)},options,cache);
+		}
 
-		inline core::smart_refctd_ptr<ICPUShader> compileToSPIRV(system::IFile* sourceFile, const SCompilerOptions& options) const
+		inline core::smart_refctd_ptr<ICPUShader> compileToSPIRV(system::IFile* sourceFile, const SCompilerOptions& options, core::smart_refctd_ptr<CCache> cache = nullptr) const
```

Comment:
dont pass arguments to functions as smart pointers that you're not going to own later after the function returns, use a raw (dumb) pointer for the `cache`